tsfn: Implement copy constructor

KevinEady opened the pull request:
Removes the `unique_ptr` from `ThreadSafeFunction`, thereby allowing
copies.

Ref: https://github.com/nodejs/node-addon-api/issues/524

KevinEady commented:
Hi @legendecas ,

So... IMO I don't think it should automatically acquire/release. I believe it adds an assumption that each copy must perform an `Acquire`, which I think is not true. In fact, I think it would actually cause problems in certain implementations. 

Also, if `Acquire()` returns `napi_closing`, I would not consider this a fatal error so we shouldn't throw in the copy ctor. How would we handle this? Storing the status on the newly constructed object for the developer to check after copy, eg. in the thread entry point? I dunno, seems clunky...

gabrielschulhof commented:
@legendecas IINM passing to another thread may not be the only reason for copying a `Napi::ThreadSafeFunction`. For example, if you used it as a key in a hash table ...

gabrielschulhof commented:
Also IINM it should be easy to subclass `Napi::ThreadSafeFunction` and add the automatic `Acquire()`/`Release()` for those that need it.

KevinEady commented:
So I have no idea about the ambiguous function calls for `New` as mentioned in https://github.com/nodejs/node-addon-api/issues/544#issuecomment-534175977 but I don't think it should block this. If so, I believe this is ready.

mhdawson commented:
I agree that the problem related to ambiguous function calls is independent and could be addressed in a follow on PR.

mhdawson commented on test/threadsafe_function/threadsafe_function_sum.cc:
```diff
@@ -0,0 +1,155 @@
+  TestData* testData = static_cast<TestData*>(info.Data());
+  ThreadSafeFunction tsfn = testData->tsfn;
+  int threadId = testData->threads.size();
+  testData->threads.push_back( thread(entryAcquire, tsfn, threadId) );
```
I wonder if it would be useful to highlight with a comment that this is where the copy constructor is being tested.  Since the main change is to add the copy constructor it would be good to make it obvious in the test that it is being used.

KevinEady replied:
Added comment

KevinEady commented:
Hey @mhdawson, can we get this landed so I can continue work on Ref/Unref as well as the incorrect `[Non]BlockingCall` with `ThreadSafeFunction(napi_threadsafe_function)` constructor

mhdawson commented:
I'm in the process of landing. It did not seem to compile with master and the older 4.8.5 compiler.  That may just be because master need gcc 6, but I want to run a CI on 8.x before landing. 

Unfortunately, the CI job is having some issues getting the Node binaries so I'll have to figure that out first.

mhdawson commented:
This seems to fail to build on 8.x with the 4.8.5 compiler level. You can see the compile failures her: https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-node-addon-api-new/nodes=ubuntu1604-64/822/console

mhdawson commented:
@KevinEady can you take a look:

```
../threadsafe_function/threadsafe_function_sum.cc: In function ‘Napi::Value {anonymous}::TestWithTSFN(const Napi::CallbackInfo&)’:
../threadsafe_function/threadsafe_function_sum.cc:51:4: error: no matching function for call to ‘{anonymous}::TestData::TestData(<brace-enclosed initializer list>)’
   });
    ^
../threadsafe_function/threadsafe_function_sum.cc:12:8: note: candidate: {anonymous}::TestData::TestData(const {anonymous}::TestData&)
 struct TestData {
        ^
../threadsafe_function/threadsafe_function_sum.cc:12:8: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘const {anonymous}::TestData&’

```

KevinEady commented:
Hey team,

So I was actually testing this on Windows, and it was failing because defining a move operator implicitly deletes the assignment operator... And, if you think about it, the `ThreadSafeFunction` wrapper only has for its member data this `napi_threadsafe_function` which behaves as a pointer: we can use the default compiler-generated functions anyways, right? 

```
test\threadsafe_function\threadsafe_function_sum.cc(90): error C2280: 'Napi::ThreadSafeFunction &Napi::ThreadSafeFunction::operator =(const Napi::ThreadSafeFunction &)': attempting to reference a deleted function [test\build\binding_noexcept.vcxproj]
  napi.h(2088): note: compiler has generated 'Napi::ThreadSafeFunction::operator =' here
  napi.h(2088): note: 'Napi::ThreadSafeFunction &Napi::ThreadSafeFunction::operator =(const Napi::ThreadSafeFunction &)': function was implicitly deleted because 'Napi::ThreadSafeFunction' has a user-defined move constructor (compiling source file ..\threadsafe_function\threadsafe_function_sum.cc)
```

I did not think it was worthwhile to explicitly create these constructors, nor state them as default in napi.h , so I just removed them. Since this is structural change, I've dismissed your review @mhdawson

mhdawson commented:
CI job to validate builds on Node 8.x: https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-node-addon-api-new/854/

mhdawson commented:
@KevinEady seems to be failures on osx and aix: : https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-node-addon-api-new/854/

KevinEady commented:
Hi @mhdawson ,

Fixed some issues in the tests that I know were problematic, but this one is really stumping me..

https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-node-addon-api-new/nodes=osx1011/855/console : 

```
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/future:1170:1: error: 'future<Napi::ThreadSafeFunction>' is unavailable: introduced in macOS 10.8
future<_Rp>::get()
```

I don't know, just looking at the node name os `osx1011` ... but is this build targeting osx 10.8 ...?

KevinEady commented:
Hi @mhdawson / @gabrielschulhof ,

So I re-wrote my tests to use `std::condition_variable` and `std::mutex` instead of promises/futures. Let me know if this is suitable now.

KevinEady commented:
Hi @legendecas ,

> I noticed there are still usages on promises/futures in `test/threadsafe_function/threadsafe_function_sum.cc`. Are they intended?

So I noticed I did not remove the <future> include, which I now replaced with the correct ones for use with condition_variable/mutex.

However, the promise / deferred you see are `Napi::Promise`s, not std library. They are used in the test to ensure proper synchronization. Does this address your concern?

mhdawson commented:
CI run to see if it works on 8.x https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-node-addon-api-new/906/

mhdawson commented:
@legendecas let us know if the latest changes address your concerns and then approve if appropriate. Will wait to hear from you before landing.

legendecas commented on napi-inl.h:
```diff
@@ -4083,34 +4070,34 @@ inline napi_status ThreadSafeFunction::NonBlockingCall(
 ThreadSafeFunction::GetContext() const {
   void* context;
-  napi_get_threadsafe_function_context(*_tsfn, &context);
+  napi_get_threadsafe_function_context(_tsfn, &context);
```
Abort if failed here?

legendecas replied:
Since the change doesn't change the behavior here, we could address this issue in another PR. Opened issue here https://github.com/nodejs/node-addon-api/issues/581.

mhdawson replied:
@legendecas I agree, can you open an issue to make sure we fix it later?

KevinEady replied:
@mhdawson the issue is above, and PR for fix (which is also conflicting with this PR) is #583

mhdawson replied:
@KevinEady - oops should have read more carefullly. Thanks.

mhdawson commented:
Landed as 2e71842f633b4138baddc2982079f4590e991fa6
